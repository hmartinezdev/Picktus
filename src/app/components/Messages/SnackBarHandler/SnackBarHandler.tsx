import React, { Component } from 'react';
import { Transition, TransitionGroup } from 'react-transition-group';
import SnackBar from '../SnackBar';
import { ISnackBarHandlerProps } from './SnackBarHandler.type';

class SnackBarHandler extends Component<ISnackBarHandlerProps> {
  public static defaultProps = {
    snackbars: [],
  };

  constructor(props: ISnackBarHandlerProps) {
    super(props);
  }

  private dismissSnackBar() {
    const { snackbars, dismissSnackBar } = this.props;
    if (snackbars.length > 0) {
      setTimeout(dismissSnackBar, 2000);
    }
  }

  public componentDidMount() {
    this.dismissSnackBar();
  }

  public componentDidUpdate(prevProps: ISnackBarHandlerProps) {
    const currentSnackBar = this.props.snackbars[0];
    const previousSnackBar = prevProps.snackbars[0];

    if (!currentSnackBar) {
      return;
    }

    if (!previousSnackBar || previousSnackBar.id !== currentSnackBar.id) {
      this.dismissSnackBar();
    }
  }

  public render(): React.ReactElement<SnackBarHandler> {
    const { snackbars } = this.props;

    return (
      <div>
        <TransitionGroup className="snackbars" appear={true}>
          {snackbars.map((value, key) => {
            if (key === 0) {
              return (
                <Transition key={value.id} timeout={700} mountOnEnter={true} unmountOnExit={true}>
                  {(status) => (
                    <div key={value.id} className={`snackbar__container snackbar__container--${status}`}>
                      <SnackBar text={value.text} level={value.level} />
                    </div>
                  )}
                </Transition>
              );
            }
          })}
        </TransitionGroup>
        <style jsx>{`
          .snackbar__container {
            position: absolute;
            bottom: 0px;
            left: 50%;
            transition: all 350ms ease-out;
            transform: translate(-50%, 120%);
            z-index: 1000;
            width: 90%;
          }

          .snackbar__container--entering {
            animation: slideUp 700ms ease-out;
          }

          .snackbar__container--entered {
            transform: translate(-50%, 0%);
          }

          .snackbar__container--entered,
          .snackbar__container--entering {
            z-index: 2;
          }

          .snackbar__container--exiting,
          .snackbar__container--exited {
            transform: translate(-50%, 120%);
            z-index: 1;
          }

          @keyframes slideUp {
            from {
              transform: translate(-50%, 120%);
            }

            to {
              transform: translate(-50%, 0%);
            }
          }
        `}</style>
      </div>
    );
  }
}

export default SnackBarHandler;
