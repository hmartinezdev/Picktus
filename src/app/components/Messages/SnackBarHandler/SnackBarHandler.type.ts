import { IPicktusError, IPicktusMessage } from '@store/reducers/message';
import { DismissSnackBar } from '@store/reducers/message/message-actions';

export interface IReduxStateProps {
  snackbars: Array<IPicktusError | IPicktusMessage>;
}

export interface IMessageHandlerState {
  snackbars: IPicktusError | IPicktusMessage | null;
}

export interface IDispatchProps {
  dismissSnackBar: () => DismissSnackBar;
}

export type ISnackBarHandlerProps = IReduxStateProps & IDispatchProps;
