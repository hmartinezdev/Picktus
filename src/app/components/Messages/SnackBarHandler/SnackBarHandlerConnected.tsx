import { IReduxState } from '@store/reducers';
import { dismissSnackBar } from '@store/reducers/message/message-actions';
import { connect } from 'react-redux';
import SnackBarHandler from './SnackBarHandler';
import { IReduxStateProps } from './SnackBarHandler.type';

const mapStateToProps = (state: IReduxState): IReduxStateProps => {
  return {
    snackbars: state.message.snackbars,
  };
};

export default connect(
  mapStateToProps,
  { dismissSnackBar }
)(SnackBarHandler);
