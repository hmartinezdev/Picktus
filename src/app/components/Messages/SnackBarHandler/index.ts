import SnackBarHandlerConnected from './SnackBarHandlerConnected';

export default SnackBarHandlerConnected;
