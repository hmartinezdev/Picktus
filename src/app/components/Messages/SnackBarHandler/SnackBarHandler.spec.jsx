import { shallow } from 'enzyme';
import React from 'react';
import SnackBarHandler from './SnackBarHandler';

let wrapper;

const setup = (props = {}) => shallow(<SnackBarHandler {...props} />);

describe('<SnackBarHandler />', () => {
  test('should render properly', () => {
    wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  test('should render with snackbars defined', () => {
    const spyDissmiss = jest.fn();
    wrapper = setup({ snackbars: [{ id: 'ok' }], dismissSnackBar: spyDissmiss });
    expect(wrapper).toMatchSnapshot();
  });

  describe('dismissSnackBar', () => {
    test('it should set a timeout to dismiss the current snackbar if there is at least a snackbar', () => {
      const spyTimeOut = jest.spyOn(window, 'setTimeout').mockImplementation((fn) => fn());
      const spyDissmiss = jest.fn();
      wrapper = setup({ snackbars: [{ id: 'ok' }], dismissSnackBar: spyDissmiss });

      expect(spyTimeOut).toHaveBeenCalledWith(spyDissmiss, 2000);
      expect(spyDissmiss).toHaveBeenCalled();
    });
  });

  describe('componentDidUpdate', () => {
    test('it should call dismissSnackBar if there is still a new snackbar to be displayed', () => {
      const spyTimeOut = jest.spyOn(window, 'setTimeout').mockImplementation((fn) => fn());
      const spyDissmiss = jest.fn();
      wrapper = setup({ snackbars: [{ id: 'ok' }, { id: 'test' }], dismissSnackBar: spyDissmiss });
      wrapper.instance().componentDidUpdate({ snackbars: [{ id: 'test' }], dismissSnackBar: spyDissmiss });

      expect(spyDissmiss).toHaveBeenCalledTimes(2);
    });
  });
});
