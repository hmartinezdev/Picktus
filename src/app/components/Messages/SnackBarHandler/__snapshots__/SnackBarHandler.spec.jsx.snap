// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SnackBarHandler /> should render properly 1`] = `
<div>
  <TransitionGroup
    appear={true}
    childFactory={[Function]}
    className="snackbars"
    component="div"
  />
  <style
    jsx={true}
  >
    
          .snackbar__container {
            position: absolute;
            bottom: 0px;
            left: 50%;
            transition: all 350ms ease-out;
            transform: translate(-50%, 120%);
            z-index: 1000;
          }

          .snackbar__container--entering {
            animation: slideUp 700ms ease-out;
          }

          .snackbar__container--entered {
            transform: translate(-50%, 0%);
          }

          .snackbar__container--entered,
          .snackbar__container--entering {
            z-index: 2;
          }

          .snackbar__container--exiting,
          .snackbar__container--exited {
            transform: translate(-50%, 120%);
            z-index: 1;
          }

          @keyframes slideUp {
            from {
              transform: translate(-50%, 120%);
            }

            to {
              transform: translate(-50%, 0%);
            }
          }
        
  </style>
</div>
`;

exports[`<SnackBarHandler /> should render with snackbars defined 1`] = `
<div>
  <TransitionGroup
    appear={true}
    childFactory={[Function]}
    className="snackbars"
    component="div"
  >
    <Transition
      appear={false}
      enter={true}
      exit={true}
      in={false}
      key="ok"
      mountOnEnter={true}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      timeout={700}
      unmountOnExit={true}
    >
      <Component />
    </Transition>
  </TransitionGroup>
  <style
    jsx={true}
  >
    
          .snackbar__container {
            position: absolute;
            bottom: 0px;
            left: 50%;
            transition: all 350ms ease-out;
            transform: translate(-50%, 120%);
            z-index: 1000;
          }

          .snackbar__container--entering {
            animation: slideUp 700ms ease-out;
          }

          .snackbar__container--entered {
            transform: translate(-50%, 0%);
          }

          .snackbar__container--entered,
          .snackbar__container--entering {
            z-index: 2;
          }

          .snackbar__container--exiting,
          .snackbar__container--exited {
            transform: translate(-50%, 120%);
            z-index: 1;
          }

          @keyframes slideUp {
            from {
              transform: translate(-50%, 120%);
            }

            to {
              transform: translate(-50%, 0%);
            }
          }
        
  </style>
</div>
`;
